### ISSUE TITLE
{word} expressions not always giving expected results

### ISSUE THREAD

**Issue by kreuzberger** [ThoSe1990/cwt-cucumber#38]
Testing the version from #31 i found some results that do not work like expected for me. Taking the box example from the example as base:

```

Feature: My first feature
  This is my cucumber-cpp hello world

  Scenario Outline: First Scenario Outline
    Given An empty box
    When I place <count> x <item> in it
    Then The box contains <count> item(s)
    Then The <count>. item is <item>

    Examples:
      | count | item      |
      | 1     | "apple"   |
      | 2     | "bananas" |

```
Adding adding an additional test at the end of the scenario outline.

# Word and String are exchangeable.
not changing the step definition `THEN(test, "The {int}. item is {string}")` the tests work as expected
Testing with 

```
    Examples:
      | count | item    |
      | 1     | apple   |
      | 2     | bananas |

```
works also, but the {string} should have contained `"`? Testing with the website (cucumber expressions) STRING should always match quoted string, while word should return the content with quotes. {string} should never match "unquoted string"

Testing BOTH examples with `THEN(test, "The {int}. item is {word}")` works, but always give back the "unquoted" string.

# Words with characters also used for integers
If the word or string operator is run on strings with integer/float characters like `, . -` the results get more or less wrong

Testing the examples above with `THEN(test, "The {int}. item is {word}")`
 Input is:

```
```

Output of the example is
```
Feature: My first feature  examples/features/2_scenario_outline.feature:2

  With Examples:
  | count | item        |
  | 1     | 100         |

  With Examples:
  | count | item        |
  | 2     | -100        |

  With Examples:
  | count | item        |
  | 3     | -100        |

  With Examples:
  | count | item        |
  | 4     | -100,0      |

  With Examples:
  | count | item        |
  | 5     | 100,0       |

  With Examples:
  | count | item        |
  | 6     | -100.000000 |

  With Examples:
  | count | item        |
  | 7     | 100.000000  |

-- Hook after all

7 Scenarios (7 passed)
28 Steps (28 passed)

```
It seems that some integer detection takes over precedence of the "word" detection
3: would expect the result with quotes
4: would expect the result with quotes
5: Would expect -100,0 as string, minus is lost
6: would expect -100.0 as string, extra 0 from conversion to double
7: same as 6

Using `THEN(test, "The {int}. item is {string}")` does not change anything int the results. Expected here is to fail 1,2,5,6,7 cause item contains now quotes?

# Spaces in box items
Testing with 
```
   Examples:
      | count | item         |
      | 1     | apple        |
      | 2     | banana       |
      | 3     | ice cream    |
      | 4     | "ice cream"  |
```
and `THEN(test, "The {int}. item is {word}")`

gives no error.
```
Scenario Outline: First Scenario Outline  examples/features/2_scenario_outline.feature:5
[   PASSED    ] Given An empty box  examples/features/2_scenario_outline.feature:6
[   PASSED    ] When I place <count> x <item> in it  examples/features/2_scenario_outline.feature:7
[   PASSED    ] Then The box contains <count> item(s)  examples/features/2_scenario_outline.feature:8
[   PASSED    ] Then The <count>. item is <item>  examples/features/2_scenario_outline.feature:9
  With Examples:
  | count | item      |
  | 3     | ice cream |

Scenario Outline: First Scenario Outline  examples/features/2_scenario_outline.feature:5
[   PASSED    ] Given An empty box  examples/features/2_scenario_outline.feature:6
[   PASSED    ] When I place <count> x <item> in it  examples/features/2_scenario_outline.feature:7
[   PASSED    ] Then The box contains <count> item(s)  examples/features/2_scenario_outline.feature:8
[   PASSED    ] Then The <count>. item is <item>  examples/features/2_scenario_outline.feature:9
  With Examples:
  | count | item      |
  | 4     | ice cream |
```
Expected here that 3,4 fails cause the {word} could not find a valid step definition?

**Comment by kreuzberger** [comment]
Tested with the code before the #37 integration....

**Comment by ThoSe1990** [comment]
Fixed https://github.com/ThoSe1990/cwt-cucumber/pull/40

**Comment by kreuzberger** [comment]
Sorry, still not working for me?

**Comment by ThoSe1990** [comment]
How does the step definition loook like? 

I just opened a branch with an example: https://github.com/ThoSe1990/cwt-cucumber/tree/word-example

you can find those two steps: 
```cpp
struct foo 
{
  std::string word;
  std::string anonymous; 
};

WHEN(word_anonymous_given, "A {word} and {}")
{
  std::string word = CUKE_ARG(1);
  cuke::context<foo>().word = word;
 
  std::string anonymous = CUKE_ARG(2);
  cuke::context<foo>().anonymous = anonymous;
}

THEN(word_anonymous_then, "They will match {string} and {string}")
{
  std::string expected_word = CUKE_ARG(1);
  std::string expected_anonymous = CUKE_ARG(2);
  
  cuke::equal(expected_word, cuke::context<foo>().word);
  cuke::equal(expected_anonymous, cuke::context<foo>().anonymous);
}
```

```
  Scenario Outline: a scenario outline  
    When A <word> and <anonymous>
    Then They will match <expected_word> and <expected_anonymous>
    
    Examples:
      | word  | anonymous | expected_word | expected_anonymous |
      | 123    | 999              | "123"                  | "999"                            | 
```

This passes and all values are treated as string

**Comment by kreuzberger** [comment]
```
diff --git a/examples/features/2_scenario_outline.feature b/examples/features/2_scenario_outline.feature
index 6ca5db3..878a66a 100644
--- a/examples/features/2_scenario_outline.feature
+++ b/examples/features/2_scenario_outline.feature
@@ -6,9 +6,12 @@ Feature: My first feature
     Given An empty box
     When I place <count> x <item> in it
     Then The box contains <count> item(s)
+    Then The <count>. item is <item>^M
 
     Examples:
-      | count | item      |
-      | 1     | "apple"   |
-      | 2     | "bananas" |
+      | count | item       |^M
+      | 1     | apple      |^M
+      | 2     | bananas    |^M
+      | 3     | -1,0       |^M
+      | 4     | ice cream  |^M
 
diff --git a/examples/step_definition.cpp b/examples/step_definition.cpp
index 144dabd..f0f3d55 100644
--- a/examples/step_definition.cpp
+++ b/examples/step_definition.cpp
@@ -56,7 +56,7 @@ WHEN(add_table_rows_hash, "I add the following item with rows_hash():")
                                  hash_rows["QUANTITY"].as<long>());
 }
 
-THEN(test, "The {int}. item is {string}")
+THEN(test, "The {int}. item is {word}")^M
 {
   const std::size_t number = CUKE_ARG(1);
   const std::size_t idx_zero_based = number - 1;
```

**Comment by kreuzberger** [comment]
from your branch i changed the outline
```
diff --git a/examples/features/2_scenario_outline.feature b/examples/features/2_scenario_outline.feature
index ab89877..16bdc86 100644
--- a/examples/features/2_scenario_outline.feature
+++ b/examples/features/2_scenario_outline.feature
@@ -19,3 +19,8 @@ Feature: My first feature
     Examples:
       | word  | anonymous | expected_word | expected_anonymous |
       | 123   | 999       | "123"         | "999"              | 
+      | -1,0  | -99       | "-1,0"        | "-99"              | ^M
+      | -1.0  | -99       | "-1.0"        | "-99"              | ^M
+      | abc   | ice cream | abc           | "ice cream"        | ^M
+
```

gives 
```
Scenario Outline: a scenario outline  ../examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A <word> and <anonymous>  ../examples/features/2_scenario_outline.feature:16
Value -1,0 is not equal to 1,0 in following step:
[   FAILED    ] Then They will match <expected_word> and <expected_anonymous>  ../examples/features/2_scenario_outline.feature:17
  With Examples:
  | word      | anonymous | expected_word | expected_anonymous |
  | 1,0       | -99       | -1,0          | -99                |

Scenario Outline: a scenario outline  ../examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A <word> and <anonymous>  ../examples/features/2_scenario_outline.feature:16
Value -1.0 is not equal to -1.000000 in following step:
[   FAILED    ] Then They will match <expected_word> and <expected_anonymous>  ../examples/features/2_scenario_outline.feature:17
  With Examples:
  | word      | anonymous | expected_word | expected_anonymous |
  | -1.000000 | -99       | -1.0          | -99                |

Scenario Outline: a scenario outline  ../examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A <word> and <anonymous>  ../examples/features/2_scenario_outline.feature:16
[   PASSED    ] Then They will match <expected_word> and <expected_anonymous>  ../examples/features/2_scenario_outline.feature:17
  With Examples:
  | word      | anonymous | expected_word | expected_anonymous |
  | abc       | ice cream | abc           | ice cream          |
```

**Comment by kreuzberger** [comment]
Using
```
   Examples:
      | word  | anonymous | expected_word | expected_anonymous |
      | a c   | ice cream | a             | "ice cream"        | 
      | 123   | -999.0    | "123"         | "-999.0"           | 
     ```

shows that the number problem is the same for {word} and {}.
Space handling in the space divided word example should lead to errors i think.

**Comment by ThoSe1990** [comment]
yh, I noticed that during implementation, so as mentioned before, the table values are parsed as double and in order to represent them as string std::to_string is called, this brings the additional decimal numbers in. 

But I just noticed that I could use the format library and this should do the trick.

**Comment by ThoSe1990** [comment]
I just opened https://github.com/ThoSe1990/cwt-cucumber/pull/42, will take a look on it tomorrow and merge it. For now it looks good I guess

**Comment by ThoSe1990** [comment]
TL;DR: I refactored a lot and came to a different approach. All tests pass and internally I hold all values as std::string. it comes down to the step definition to which type you want to convert. 

---

In earlier implementations I needed the type information, since i parsed features/steps myself. In order to match to the correct type, I used type erasure in `cuke::value`. So every value was represented as `cuke::value` and with some type traits I was able to detect if `{int}` matches an `int`, and so on and so forth. 

Since this all is now out and we introduced regex to find the steps, I could remove the interface inside `cuke::value` and every value is now represented as string internally. Only in the finale state when the developer in their step definition access the value, i convert the strings to  int, double, etc. 

With this the printing issue resolves, because there is no conversion from string -> float/double and back to string again. (even though it looked good with the format library) 

This is a massive simplification internally and for now all tests pass and it looks very good.  I'll play around and sleep a night on it but this looks as a great improvement to me. 

So at this point, thank you so much, for having a detailed look on this project, it improved already a lot. 

If there aren't any issues, I'll merge this branch.

**Comment by ThoSe1990** [comment]
done ✅ https://github.com/ThoSe1990/cwt-cucumber/pull/42

**Comment by kreuzberger** [comment]
Works and could use same feature files from other cucumber implementations. :+1:
